src/storage/compression/bitpacking.cpp

```diff
@@ -0,0 +1,599 @@
+	}
+}
+
+template <class T, class PRE_CAST_TYPE>
+void BitpackingScan(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result) {
+	BitpackingScanPartial<T, PRE_CAST_TYPE>(segment, state, scan_count, result, 0);
+}
+
+//===--------------------------------------------------------------------===//
+// Fetch
+//===--------------------------------------------------------------------===//
+template <class T, class PRE_CAST_TYPE>
+void BitpackingFetchRow(ColumnSegment &segment, ColumnFetchState &state, row_t row_id, Vector &result,
+                        idx_t result_idx) {
+	BitpackingScanState<T, PRE_CAST_TYPE> scan_state(segment);
+	scan_state.Skip(segment, row_id);
+	auto result_data = FlatVector::GetData<T>(result);
+	T *current_result_ptr = result_data + result_idx;
+
+	// Todo clean up, is reused in partialscan
+	idx_t offset_in_compression_group =
+	    scan_state.position_in_group % BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE;
+
+	data_ptr_t decompression_group_start_pointer =
+	    scan_state.current_width_group_ptr +
+	    (scan_state.position_in_group - offset_in_compression_group) * scan_state.current_width / 8;
+
+	scan_state.decompress_function((data_ptr_t)scan_state.decompression_buffer, decompression_group_start_pointer,
+	                               scan_state.current_width);
+
+	if (std::is_same<T, PRE_CAST_TYPE>::value) {
+		*current_result_ptr = *(T *)(scan_state.decompression_buffer + offset_in_compression_group);
+	} else {
+		*current_result_ptr = (T) * (PRE_CAST_TYPE *)(scan_state.decompression_buffer + offset_in_compression_group);
+	}
+}
+template <class T, class PRE_CAST_TYPE>
+void BitpackingSkip(ColumnSegment &segment, ColumnScanState &state, idx_t skip_count) {
+	auto &scan_state = (BitpackingScanState<T, PRE_CAST_TYPE> &)*state.scan_state;
+	scan_state.Skip(segment, skip_count);
+}
+
+//===--------------------------------------------------------------------===//
+// Get Function
+//===--------------------------------------------------------------------===//
+template <class T, class PRE_CAST_TYPE = T>
+CompressionFunction GetBitpackingFunction(PhysicalType data_type) {
+	return CompressionFunction(CompressionType::COMPRESSION_BITPACKING, data_type,
+	                           BitpackingInitAnalyze<T, PRE_CAST_TYPE>, BitpackingAnalyze<T, PRE_CAST_TYPE>,
+	                           BitpackingFinalAnalyze<T, PRE_CAST_TYPE>, BitpackingInitCompression<T, PRE_CAST_TYPE>,
+	                           BitpackingCompress<T, PRE_CAST_TYPE>, BitpackingFinalizeCompress<T, PRE_CAST_TYPE>,
+	                           BitpackingInitScan<T, PRE_CAST_TYPE>, BitpackingScan<T, PRE_CAST_TYPE>,
+	                           BitpackingScanPartial<T, PRE_CAST_TYPE>, BitpackingFetchRow<T, PRE_CAST_TYPE>,
+	                           BitpackingSkip<T, PRE_CAST_TYPE>);
+}
+
+CompressionFunction BitpackingFun::GetFunction(PhysicalType type) {
+	switch (type) {
+	case PhysicalType::INT8:
```

Comment:
Could this work for `PhysicalType::BOOL` as well?

Reply:
Yes it would!